server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -180,56 +170,41 @@ public ClusterState updateState(ClusterState currentState,
                 * the node?
                 */
-            routingTableBuilder.updateNumberOfReplicas(updatedNumberOfReplicas, actualIndices);
-            metadataBuilder.updateNumberOfReplicas(updatedNumberOfReplicas, actualIndices);
+            routingTableBuilder.updateNumberOfReplicas(updatedNumberOfReplicas, indexes);
+            metadataBuilder.updateNumberOfReplicas(updatedNumberOfReplicas, indexes);
             LOGGER.info("updating number_of_replicas to [{}] for indices {}", updatedNumberOfReplicas,
-                    actualIndices);
+                    partitions);
         }
 
         ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
-        maybeUpdateClusterBlock(actualIndices, blocks, IndexMetadata.INDEX_READ_ONLY_BLOCK,
+        maybeUpdateClusterBlock(indexes, blocks, IndexMetadata.INDEX_READ_ONLY_BLOCK,
                 IndexMetadata.INDEX_READ_ONLY_SETTING, openSettings);
-        maybeUpdateClusterBlock(actualIndices, blocks, IndexMetadata.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK,
+        maybeUpdateClusterBlock(indexes, blocks, IndexMetadata.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK,
                 IndexMetadata.INDEX_BLOCKS_READ_ONLY_ALLOW_DELETE_SETTING, openSettings);
-        maybeUpdateClusterBlock(actualIndices, blocks, IndexMetadata.INDEX_METADATA_BLOCK,
+        maybeUpdateClusterBlock(indexes, blocks, IndexMetadata.INDEX_METADATA_BLOCK,
                 IndexMetadata.INDEX_BLOCKS_METADATA_SETTING, openSettings);
-        maybeUpdateClusterBlock(actualIndices, blocks, IndexMetadata.INDEX_WRITE_BLOCK,
+        maybeUpdateClusterBlock(indexes, blocks, IndexMetadata.INDEX_WRITE_BLOCK,
                 IndexMetadata.INDEX_BLOCKS_WRITE_SETTING, openSettings);
-        maybeUpdateClusterBlock(actualIndices, blocks, IndexMetadata.INDEX_READ_BLOCK,
+        maybeUpdateClusterBlock(indexes, blocks, IndexMetadata.INDEX_READ_BLOCK,
                 IndexMetadata.INDEX_BLOCKS_READ_SETTING, openSettings);
 
-        for (Index index : openIndices) {
-            IndexMetadata indexMetadata = metadataBuilder.getSafe(index);
+        for (IndexMetadata indexMetadata : indexes) {
             Settings.Builder updates = Settings.builder();
             Settings.Builder indexSettings = Settings.builder().put(indexMetadata.getSettings());
-            if (indexScopedSettings.updateDynamicSettings(openSettings, indexSettings, updates, index.getName())) {
+            Settings toUpdate = indexMetadata.getState() == IndexMetadata.State.OPEN ? openSettings : closedSettings;
+            String indexName = indexMetadata.getIndex().getName();
+            // TODO can we improve updateDynamicSettings so that it returns false if the update
+            // doesn't actually change the value of the setting?  Then we can remove the `same` call
+            if (indexScopedSettings.updateDynamicSettings(toUpdate, indexSettings, updates, indexName)) {
```

Comment:
Before only open indices were changed via `updateDynamicSettings` and for closed it used `indexScopedSettings.updateSettings`. Shouldn't that difference be preserved?